gen/flights/flightplan.py

```diff
@@ -114,6 +116,65 @@ def tot_waypoint(self) -> Optional[FlightWaypoint]:
         failed to generate. Nevertheless, we have to defend against it.
         """
         raise NotImplementedError
+    
+    @cached_property
+    def bingo_fuel(self) -> int:
+        """Bingo fuel value for the FlightPlan
+        """
+        if self.waypoints is None or self.farthest_wpt_from_arrival is None:
+            return 0
+
+        distanceToArrival = self.farthest_wpt_from_arrival.position.distance_to_point(self.flight.arrival.position)
+        distanceToArrival = meter_to_nm(distanceToArrival)
+
+        bingo = 1000 # Minimum Emergency Fuel
+        bingo += 500 # Visual Traffic
+        bingo += 15 * distanceToArrival
+
+        if self.flight.divert is not None and self.farthest_wpt_from_divert is not None:
+            assert self.farthest_wpt_from_divert is not None
+            assert self.flight.divert is not None
+            distanceToDivert = self.farthest_wpt_from_divert.position.distance_to_point(self.flight.divert.position)
+            distanceToDivert = meter_to_nm(distanceToDivert)
+            bingo += 10 * distanceToDivert
+
+        return round(bingo / 100) * 100
+
+    @cached_property
+    def joker_fuel(self) -> int:
+        return self.bingo_fuel + 1000
+
+    @property
+    def farthest_wpt_from_arrival(self) -> Optional[FlightWaypoint]:
+        """Returns the farthest waypoint of the flight plan from the arrival airport.
+        """
+        if self.flight.arrival is None:
+            return None
+
+        return self.farthest_wpt_from(self.flight.arrival)
+
+    @property
+    def farthest_wpt_from_divert(self) -> Optional[FlightWaypoint]:
+        """Returns the farthest waypoint of the flight plan from the divert airport.
+        """
+        if self.flight.divert is None:
+            return None
+ 
+        return self.farthest_wpt_from(self.flight.divert)
+    
+    def farthest_wpt_from(self, cp) -> Optional[FlightWaypoint]:
```

Comment:
You're never actually using the waypoint that this returns, and instead are recalculating the distance at the call site. Instead, how about:

```
def max_distance_from(self, cp: ControlPoint) -> int:
    if not self.waypoints:
        return 0
    return max([cp.position.distance_to_point(w.position) for w in self.waypoints]
```

(or return `None` in the no waypoints case, not sure at a glance which is easier to work with)